test/new-e2e/system-probe/system-probe-test-env.go

```diff
@@ -188,14 +185,15 @@ func NewTestEnv(name, x86InstanceType, armInstanceType string, opts *SystemProbe
 	}
 
 	var upResult auto.UpResult
+	var az string
 	ctx := context.Background()
 	currentAZ := 0 // PrimaryAZ
 	b := retry.NewConstant(3 * time.Second)
 	// Retry 4 times. This allows us to cycle through all AZs, and handle libvirt
 	// connection issues in the worst case.
 	b = retry.WithMaxRetries(4, b)
 	if retryErr := retry.Do(ctx, b, func(_ context.Context) error {
-		if az := getAvailabilityZone(opts.InfraEnv, currentAZ); az != "" {
+		if az, currentAZ = getAvailabilityZone(opts.InfraEnv, currentAZ); az != "" {
 			config["ddinfra:aws/defaultSubnets"] = auto.ConfigValue{Value: az}
 		}
```

Comment:
I don't think you need to change the function, but can do the following. That makes it much clearer what is happening to the local variable
```suggestion
		if az = getAvailabilityZone(opts.InfraEnv, currentAZ); az != "" {
			config["ddinfra:aws/defaultSubnets"] = auto.ConfigValue{Value: az}
			currentAZ++
		} else {
			currentAZ = 0		
		}
```

Reply:
I realize now that this logic is incorrect. We do not want to change the AZ when the error type does not require it. Changing it unnecessarily would cause the whole stack to start from the beginning.